### PR TITLE
Added NaNInterface class

### LINKED ISSUE #12234: Add interface for fluid properties to choose between quiet and loud NaNs
## Rationale

There are currently some fluid properties classes, such as some in IAPWS95, that emit quiet NaNs. The intention is that evaluation of fluid properties in invalid states should not necessarily terminate a simulation, just allow time step size to be cut, even when trapping NaNs in debug mode, as other NaNs might be the objective of such traps. However, there are also cases when it is desired to terminate with these NaNs as well. Therefore, it would be useful to have an option for this. Note that there might also be other cases (not just fluid properties) where this option would be useful.

## Description

The current idea to solve this issue is to create an interface class, for example, `NaNInterface` that has a boolean parameter `emit_trappable_nans` and a function `getNaN()`, which returns either a loud (and trappable) NaN or a quite NaN, based on the value of `emit_trappable_nans`. 

## Impact

New interface class that fluid properties classes could use, for example.

### PR DESCRIPTION
**Pull request by joshuahansel**
Closes #12234

### CODE REVIEW

File: modules/fluid_properties/src/base/NaNInterface.C [hunk 1]
```diff
@@ -0,0 +1,49 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NaNInterface.h"
```
Comment (permcody) [inline modules/fluid_properties/src/base/NaNInterface.C]: You might consider putting these files in an "interfaces" folder so they mirror the framework.

Comment (joshuahansel) [reply modules/fluid_properties/src/base/NaNInterface.C]: Will do.

**Comment by joshuahansel** [comment]
Is the debug build warning/failure a compiler bug? Seems like BS to me. Do I need to make some dummy implementation of the overloaded function to shut up the compiler?

**Comment by joshuahansel** [comment]
The Mastodon failure appears unrelated.

**Comment by dschwen** [comment]
The warning is not bogus. Your new function is an overload rather than a virtual method. You need to remove the `override` key word and need to realize that this method will _not_ be called on base class references and pointers. It is a new and separate method!

**Comment by joshuahansel** [comment]
I realize the compiler is *trying* to overload, but it shouldn't be - it should be overriding. The base class does have the interface I'm using:
```
Real p_from_v_e_(Real, Real) const
```
The thing is, it also has
```
void p_from_v_e(Real, Real, Real &, Real &, Real &) const
```
I spoke to David, and apparently this has happened other times recently: I do need to have some dummy implementation for the void `p_from_v_e` so that the compiler doesn't try to do the wrong thing here.

**Comment by andrsd** [comment]
Yeah. And, this only happens with gcc, clang is happy with the way @joshuahansel had it.

File: modules/fluid_properties/test/src/userobjects/NaNInterfaceTestFluidProperties.C [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  return params;
+}
+
+NaNInterfaceTestFluidProperties::NaNInterfaceTestFluidProperties(const InputParameters & parameters)
+  : SinglePhaseFluidProperties(parameters), NaNInterface(this)
+{
+}
+
+NaNInterfaceTestFluidProperties::~NaNInterfaceTestFluidProperties() {}
```
Comment (permcody) [inline modules/fluid_properties/test/src/userobjects/NaNInterfaceTestFluidProperties.C]: If you aren't defining a new base class just leave out the empty destructor. We only need to define these when we want to declare them as virtual. If you do need to define them for some reason (rare), you can just use "= default;"

**Review by andrsd — APPROVED** [review]

**Pull request merged**